### PR TITLE
Created a configurable memory cache

### PR DESCRIPTION
**Pull request by kichalla**
@SajayAntony @martinpf @Tratcher

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <packageSources>
-    <add key="AspNetVNext" value="https://www.myget.org/F/aspnetcidev/api/v3/index.json" />
+    <add key="AspNetVNext" value="https://www.myget.org/F/aspnetvolatiledev/api/v3/index.json" />
```
Comment (kichalla) [inline NuGet.config]: ignore this change

**Comment by SajayAntony** [comment]
Ignoring the nuget.config fix 
:shipit:

**Comment by Tratcher** [comment]
Why??

**Comment by kichalla** [comment]
@Tratcher For Perf scenarios

**Comment by Tratcher** [comment]
:-1: Just implement the interface and add it to DI, there's no need to change all the callers.

**Comment by kichalla** [comment]
@Tratcher I initially thought of doing so but then thought that it would affect any of our framework components which depend on `IMemoryCache` and which I believe is not what the Perf team wants.

@SajayAntony, can you confirm? When the cache needs to be disabled, its only for the user code that needs to be disabled and not the framework code itself, right?

For example, the framework depends on `IMemoryCache` service in the following places and even they would be affected. 
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs#L42
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs#L103

**Comment by SajayAntony** [comment]
@kichalla yes we only care about disabling memory cache for usercode since if that is being used in the framework then we want to make sure any issues are uncovered from framework misbehaving.

**Comment by Tratcher** [comment]
This still seems like a very strange abstraction. AppSettings.StoreInCache doesn't impact IMemoryCache.Get or Remove, just Set. Get rid of ConfigurableMemoryCache and inject AppSettings into StoreManagerController and check StoreInCache on line 56. Much less invasive.

File: src/MusicStore/Properties/AppSettings.cs [hunk 2]
```diff
@@ -3,5 +3,7 @@
     public class AppSettings
     {
         public string SiteTitle { get; set; }
+
+        public bool StoreInCache { get; set; }
```
Comment (Tratcher) [inline src/MusicStore/Properties/AppSettings.cs]: Can this default to true and be disabled via config?

Comment (Tratcher) [inline src/MusicStore/Properties/AppSettings.cs]: Naming. Store what in cache? `CacheDbLookups`?

**Comment by kichalla** [comment]
> This still seems like a very strange abstraction. AppSettings.StoreInCache doesn't impact IMemoryCache.Get or Remove, just Set. Get rid of ConfigurableMemoryCache and inject AppSettings into StoreManagerController and check StoreInCache on line 56. Much less invasive.

But this seems more invasive. This kind of check needs to be made in all places where the cache is being used. Currently we hide this behind the `ConfigurableMemoryCache` layer.

File: src/MusicStore/Properties/AppSettings.cs (hunk 2 above)
Comment (kichalla) [inline src/MusicStore/Properties/AppSettings.cs]: OK. Probably `UseCache` instead?

**Comment by Tratcher** [comment]
It's actually less invasive as you only need to do it where you call Set, three places, as opposed to the 12 places you had to change IMemoryCache to your custom type.

File: src/MusicStore/Properties/AppSettings.cs [hunk 3]
```diff
@@ -3,5 +3,7 @@
     public class AppSettings
     {
         public string SiteTitle { get; set; }
+
+        public bool UseCache { get; set; }
```
Comment (Tratcher) [inline src/MusicStore/Properties/AppSettings.cs]: :-1: Meh, it still doesn't say what you're using cache for.

**Comment by kichalla** [comment]
> It's actually less invasive as you only need to do it where you call Set, three places, as opposed to the 12 places you had to change IMemoryCache to your custom type.

Not sure I like it. Also considering what @SajayAntony mentioned earlier about isolating cache usage to catch issues any framework related issues, then we need to avoid using caching in all (get, set, remove) the cases when it is disabled.

**Comment by Tratcher** [comment]
If you never call Set for a particular key you shouldn't ever run into issues with Get or Remove for that key. @SajayAntony?

**Comment by kichalla** [comment]
ping @SajayAntony

**Comment by SajayAntony** [comment]
@Tratcher @kichalla as long as we don't change framework (MVC/EFs) caching behavior by changing the DI container I'm fine either way. The only change I think we are looking to do is disable data level caching so that we execute the entire code path on every request.

**Comment by Tratcher** [comment]
OK. @kichalla it should be enough to optionally avoid calling Set. You don't need a special cache implementation.

**Comment by kichalla** [comment]
updated @SajayAntony @Tratcher

File: src/MusicStore/Controllers/HomeController.cs [hunk 4]
```diff
@@ -26,14 +33,17 @@ public class HomeController : Controller
...
                 {
-                    // Refresh it every 10 minutes.
-                    // Let this be the last item to be removed by cache if cache GC kicks in.
-                    cache.Set(
+                    if (_appSettings.StoreInCache)
+                    {
+                        // Refresh it every 10 minutes.
+                        // Let this be the last item to be removed by cache if cache GC kicks in.
+                        cache.Set(
                         cacheKey,
```
Comment (Tratcher) [inline src/MusicStore/Controllers/HomeController.cs]: indentation?

File: src/MusicStore/Properties/AppSettings.cs [hunk 5]
```diff
@@ -3,5 +3,7 @@
     public class AppSettings
     {
         public string SiteTitle { get; set; }
+
+        public bool StoreInCache { get; set; } = true;
```
Comment (Tratcher) [inline src/MusicStore/Properties/AppSettings.cs]: Name still needs work. Store what in cache?

**Comment by Tratcher** [comment]
:shipit:

File: src/MusicStore/Properties/AppSettings.cs (hunk 5 above)
Comment (kichalla) [inline src/MusicStore/Properties/AppSettings.cs]: suggestions?

Comment (Tratcher) [inline src/MusicStore/Properties/AppSettings.cs]: CacheDbResults?

**Pull request merged**